### PR TITLE
React router [WIP]

### PR DESCRIPTION
**Pull request by ramyanaga**

### CODE REVIEW

**Comment by freiksenet** [comment]
I wonder if you could please configure prettier in your editor. There are a lot of diffs in this PR that are caused by eg wrong indentation, making it really hard to review. Thanks a lot!

File: frontend/Header/UserBar.js [hunk 1]
```diff
@@ -36,6 +37,7 @@ export default class UserBar extends Component {
   };
 
   handleOpenUserDropdown = (e: Event) => {
+    console.log('e: ', e);
```
Comment (freiksenet) [inline frontend/Header/UserBar.js]: [nit] debug console log

File: frontend/ListPad.js [hunk 2]
```diff
@@ -57,11 +58,14 @@ export default class ListPad extends Component {
   }
 
   render() {
+    console.log('pad list props', this.props);
```
Comment (freiksenet) [inline frontend/ListPad.js]: [nit] debug console log

File: frontend/PadContainer.js [hunk 3]
```diff
@@ -189,18 +189,7 @@ class PadContainer extends Component {
...
-      document.location.pathname === '/new'
-    ) {
-      window.history.pushState(
-        {},
-        'Apollo Launchpad',
-        `/${result.data.pad.id}`,
-      );
-    }
-
+    console.log('in afterDeploy');
```
Comment (freiksenet) [inline frontend/PadContainer.js]: This should use react-router to redirect to correct url.

File: frontend/index.js [hunk 4]
```diff
@@ -1,20 +1,78 @@
...
+});
+
+const getPadContainer = function() {
+  return <PadContainer id={null} />;
+};
+
+render(
+  // don't need to pass down history in react-router v4 according to stack overflow??
+  <ApolloProvider client={apolloClient}>
+      <BrowserRouter>
```
Comment (freiksenet) [inline frontend/index.js]: [nit] wrong indentation

File: frontend/index.js [hunk 5]
```diff
@@ -1,20 +1,78 @@
...
+      next();
+    },
+  },
+]);
+
+const apolloClient = new ApolloClient({
+  networkInterface,
+});
+
+const getPadContainer = function() {
```
Comment (freiksenet) [inline frontend/index.js]: [nit] This function is no longer used.

File: frontend/Header/Header.js [hunk 6]
```diff
@@ -106,8 +107,8 @@ export default class Header extends Component {
           <HeaderLogo />
 
           <div className="Management-Buttons">
-            <HeaderButton onClick={() => (document.location.href = `/new`)}>
-              New
+            <HeaderButton>
+              <Link to="/new">New</Link>
```
Comment (freiksenet) [inline frontend/Header/Header.js]: I think you can now add `to` prop HeaderButton directly.

Comment (ramyanaga) [inline frontend/Header/Header.js]: I think Danielle said using the onClick was better

Comment (stubailo) [inline frontend/Header/Header.js]: Yep, agreed!

File: frontend/Header/HeaderButton.js [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 import React, { Component } from 'react';
 import type { Children } from 'react';
+import { Link } from 'react-router-dom';
 import './HeaderButton.less';
 
 export default class HeaderButton extends Component {
```
Comment (stubailo) [inline frontend/Header/HeaderButton.js]: We need to update the component prop types to say that there is an optional `to` prop.

File: frontend/ListPad.js [hunk 8]
```diff
@@ -61,7 +62,9 @@ export default class ListPad extends Component {
     const padUser = pad.user || {};
     return (
       <div className="ListPad">
-        <a className="ListPad-Link" href={`/${this.props.pad.id}`}>
+        <Link className="ListPad-Link" to={`${this.props.pad.id}`}>
+          Pads you created
+          {/*<a className="ListPad-Link" href={`/${this.props.pad.id}`}>*/}
```
Comment (stubailo) [inline frontend/ListPad.js]: rip

File: frontend/PadContainer.js (hunk 3 above)
Comment (stubailo) [inline frontend/PadContainer.js]: Agreed!

File: frontend/index.js [hunk 9]
```diff
@@ -1,20 +1,74 @@
...
+import React, { Component } from 'react';
+import ReactDOM, { render } from 'react-dom';
+// might need to do import BrowserRouter as Router in below line (instead of just import Router)
+import { Switch, Router, Route, Link, BrowserRouter } from 'react-router-dom';
+//import App from './App';
 import './index.less';
+import PadSplit from './PadSplit';
+import PadContainer from './PadContainer';
+import ListContainer from './ListContainer';
+import history from './history';
```
Comment (stubailo) [inline frontend/index.js]: Can we set `no-unused-vars` to be an error instead of a warning?

File: frontend/index.js [hunk 10]
```diff
@@ -1,20 +1,74 @@
...
+    },
+  },
+]);
+
+const apolloClient = new ApolloClient({
+  networkInterface,
+});
+
+render(
+  // don't need to pass down history in react-router v4 according to stack overflow??
```
Comment (stubailo) [inline frontend/index.js]: womp let's not commit comments like this

File: frontend/PadContainer.js [hunk 11]
```diff
@@ -189,18 +189,6 @@ class PadContainer extends Component {
   }
 
   afterDeploy(target, result) {
-    if (
```
Comment (stubailo) [inline frontend/PadContainer.js]: Can we verify that setting the URL after save is still working? Looks like this was just deleted outright.

File: frontend/__tests__/ListPad.test.js [hunk 12]
```diff
@@ -1,42 +1,47 @@
...
-            version: '0.8.9',
-          },
-        ],
-        defaultQuery: '{ hello }',
-        defaultVariables: null,
-        token: 'test-token-2',
-        draft: null,
-      }}
-    />,
+    <MemoryRouter>
```
Comment (freiksenet) [inline frontend/__tests__/ListPad.test.js]: Why do we need Router here? We want to test ListPad, not router, does ListPad depend on router somehow?

File: frontend/__tests__/UserBar.test.js [hunk 13]
```diff
@@ -13,15 +14,19 @@ expect.addSnapshotSerializer(serializer);
...
-    <UserBar
-      user={{
-        id: 'test-id',
-        githubUsername: 'example',
-        pads: [],
-      }}
-      onLogin={() => {}}
-      onLogout={() => {}}
-    />,
+    <HashRouter>
```
Comment (freiksenet) [inline frontend/__tests__/UserBar.test.js]: Same here about Router.

File: frontend/__tests__/ListPad.test.js (hunk 12 above)
Comment (ramyanaga) [inline frontend/__tests__/ListPad.test.js]: The ListPad test throws a can't find history (or smthn like that) error and fails when we don't wrap it in a MemoryRouter. For reference: https://github.com/ReactTraining/react-router/issues/4337

File: frontend/PadContainer.js (hunk 11 above)
Comment (ramyanaga) [inline frontend/PadContainer.js]: It seems to be working still (I think react-router now handles history?).

Comment (ramyanaga) [inline frontend/PadContainer.js]: actually fixed this now

**Pull request merged**